first draft of changing the issue/pr section (#279)
carpentries-incubator/better-research-software: episodes/08-open-collaboration.md

Let's edit add the following function to our code.

Copy and paste this function to add it to your code

```python
def sum_duration_by_astronaut(df, output_csv):
    """
    Summarize the duration data by each astronaut and saves resulting table to a CSV file

    Args: 
        df (pd.DataFrame): The input dataframe to be summarized
        output_csv (str): Path to save the output csv of the table generated

    
    Returns:
        sum_by_astro (pd.DataFrame): Data frame with a row for each astronaut and a summarized column 
    """
    subset = df.loc[:,['crew', 'duration']] # subset to work with only relevant columns
    subset = add_duration_hours_variable(subset) # need duration_hours for easier calcs
    subset = subset.drop('duration', axis=1) # dropping extra duration file as those don't calculate correctly
    subset = subset.groupby('crew').sum() 
    print(f'Saving to CSV file {output_csv}')
    subset.to_csv(output_csv) # writing new table to specified location
    return subset
```

Then add the following to your main function, after the `eva_data` variable is created

```python
dur_by_astro = sum_duration_by_astronaut(eva_data, dur_by_astro_csv)
```

> And we will add this line in our code to run if main, before or after we define the `graph_file` variable.
> ```python

Reviewer: ```suggestion
And we will add this line in our code to run if main, before or after we define the `graph_file` variable.

```python
```

Adding an empty line to be on the safe side.

dur_by_astro_csv = 'results/duration_by_astronaut.csv'
```

> Now, lets test run our script, make sure we don't get any errors.

Reviewer: ```suggestion
Now, lets test run our script, to make sure we don't get any errors.
```

```bash
(venv_spacewalks) $ python3 eva_data_analysis.py
```

> 

Reviewer: I feel like a sentence is missing here, about verifying the change we made. Perhaps you could add `ls results` to the code block above and include the output? The presence of the CSV file would serve as enough proof that the function was run, without looking at the actual content of the file and giving the game away about the bug we want to report later on.

Now let's add and commit the new version of the code to our `02-sum-by-astro-feat` branch.
First we will check that we are on the right branch using either `git branch` or `git status`.

In our previous issue, we closed the issue manually by clicking the "Close" button in the GitHub web interface.
We can also use keywords such "fixes", "fixed", "close", "closed" or "closes" followed by a # symbol and the issue number and it will automatically close the issue when the code is accepted(merged) into main.
We will also add this to our commit message.

### Closing an issue

Once an issue is solved then it can be closed. 

> This can be done either by pressing the "Close" button in the GitHub web interface or by making a commit which includes the word "fixes", "fixed", "close", "closed" or "closes" followed by a # symbol and the issue number.

Reviewer: There is some duplication here. Can this be condensed down to a single explanation?

(venv_spacewalks) $ git switch main
```

And if we go back to the `02-sum-by-astro-feat` branch it will re-appear.

```bash
(venv_spacewalks) $ git switch 02-sum-by-astro-feat
```

If we want to push our changes to a remote such as GitHub we have to tell the `git push` command which branch to push to. If the branch doesn't exist on the remote (as it currently won't)
then it will be created. 

```bash
(venv_spacewalks) $ git push origin 02-sum-by-astro-feat
```

If we now refresh the GitHub webpage for this repository we should see the `02-sum-by-astro-feat` branch has appeared in the list of branches.

::: spoiler 

#### How to pull changes from a remote branch 
If we needed to pull changes from a branch on a remote 
(for example if we have made changes on another computer or via GitHub's web based editor), 
then we can specify a branch on a `git pull` command.

```bash
git pull origin 02-sum-by-astro-feat
```
::::::::::

:::: spoiler

#### Merging branches

> If we are working alone, when we have completed working on a branch (for example adding a feature or fixing a bug) then we might merge our branch back

Reviewer: ```suggestion
If we are working alone, when we have completed working on a branch (for example adding a feature or fixing a bug) then we can merge our branch back
```

I feel that "might" is too ambiguous. If we have made the changes and we are happy with them, why would we not merge them?

Now we are back on the main branch we can go ahead and merge the changes from the bugfix branch:

```bash
(venv_spacewalks) $ git merge 02-sum-by-astro-feat
```
::::::::::::

### Pull requests

> On larger projects, with collaborators, we might need to have a code review process before changes are merged, especially before they are merged onto the main branch that might be what is being released as the public version of the software. 

Reviewer: ```suggestion
On larger projects, with collaborators, we might include a code review process before changes are merged. A quality control step like this is especially important if the main branch of the project is being released as the public version of the software. 
```

3. Rebase and merge

The default is option 1, which will keep all of the commits made on our branch intact. 
This can be useful for seeing the whole history of our work, but if we've done a lot of minor edits or attempts at creating the feature it can be excessive to have all of this history saved. 

> This is where the second option comes in, this will place all of our changes from the branch into just a single commit, this might be much more obvious to other developers who will now see our feautre addition as a single commit in the history. 

Reviewer: ```suggestion
This is where the second option comes in, this will place all of our changes from the branch into just a single commit, this might be much more obvious to other developers who will now see our feature addition as a single commit in the history. 
```

GitHub will track that a fork has been made displays a "Contribute" button to create a pull request back to the original repository. 
Using this we can request that the changes on our fork are incorporated by the upstream project.

::::::::

:::  challenge

### Practice with Issues and PRs (Pull Requests)

We have a bug in our code!  If we look at the results in `results/duration_by_astronaut.csv`, the crew column has groups of crew and we wanted to calculate this per astronaut. 

1. Create an issue in GitHub to report this bug. A good issue description for a bug should include:

 - What the problem is, including any error messages that are displayed.
 - What version of the software it occurred with.
 - Any relevant information about the system running it, for example the operating system being used.
 - Versions of any dependent libraries.

>  - How to reproduce it.

Reviewer: Sometimes people ask for a description of the expected behaviour as well, e.g. if there isn't an error message but the user thinks the result is not correct. Would this be too much detail for our lesson?
